src/cascadia/TerminalControl/ControlCore.cpp

```diff
@@ -0,0 +1,1349 @@
+    //   handle.
+    // Arguments:
+    // - <none>
+    // Return Value:
+    // - <none>
+    winrt::fire_and_forget ControlCore::_asyncCloseConnection()
+    {
+        if (auto localConnection{ std::exchange(_connection, nullptr) })
+        {
+            // Close the connection on the background thread.
+            co_await winrt::resume_background();
+            localConnection.Close();
+            // connection is destroyed.
+        }
+    }
+
+    void ControlCore::Close()
+    {
+        if (!_closing.exchange(true))
+        {
+            // Stop accepting new output and state changes before we disconnect everything.
+            _connection.TerminalOutput(_connectionOutputEventToken);
+            _connectionStateChangedRevoker.revoke();
+
+            // GH#1996 - Close the connection asynchronously on a background
+            // thread.
+            // Since TermControl::Close is only ever triggered by the UI, we
+            // don't really care to wait for the connection to be completely
+            // closed. We can just do it whenever.
+            _asyncCloseConnection();
+
+            {
+                // GH#8734:
+                // We lock the terminal here to make sure it isn't still being
+                // used in the connection thread before we destroy the renderer.
+                // However, we must unlock it again prior to triggering the
+                // teardown, to avoid the render thread being deadlocked. The
+                // renderer may be waiting to acquire the terminal lock, while
+                // we're waiting for the renderer to finish.
+                auto lock = _terminal->LockForWriting();
+            }
+
+            if (auto localRenderEngine{ std::exchange(_renderEngine, nullptr) })
+            {
+                if (auto localRenderer{ std::exchange(_renderer, nullptr) })
+                {
+                    localRenderer->TriggerTeardown();
+                    // renderer is destroyed
+                }
+                // renderEngine is destroyed
+            }
+
+            // we don't destroy _terminal here; it now has the same lifetime as the
+            // control.
+        }
+    }
+
+    IDXGISwapChain1* ControlCore::GetSwapChain() const
+    {
+        return _renderEngine->GetSwapChain().Get();
+    }
+
+    void ControlCore::_rendererWarning(const HRESULT hr)
+    {
+        auto args{ winrt::make_self<RendererWarningArgs>(hr) };
+        _RendererWarningHandlers(*this, *args);
+    }
+
+    void ControlCore::_renderEngineSwapChainChanged()
+    {
+        _SwapChainChangedHandlers(*this, nullptr);
+    }
+
+    void ControlCore::BlinkAttributeTick()
+    {
+        auto lock = _terminal->LockForWriting();
+
+        auto& renderTarget = *_renderer;
+        auto& blinkingState = _terminal->GetBlinkingState();
+        blinkingState.ToggleBlinkingRendition(renderTarget);
+    }
+
+    void ControlCore::BlinkCursor()
+    {

```

Comment:
hmm this one doesn't need the write lock?

Reply:
(if it didn't have one before, that is a reasonable explanation. comment may hlep tho)

Reply:
It didn't have one before, but by all accounts, it should?

Reply:
Oh derp, `SetCursorOn` will lock for you.